pkg/util/informermanager/informermanager.go

```diff
@@ -0,0 +1,306 @@
+}
+
+func NewInformerManager(client dynamic.Interface, ftcInformer fedcorev1a1informers.FederatedTypeConfigInformer) InformerManager {
+	manager := &informerManager{
+		lock:                      sync.RWMutex{},
+		started:                   false,
+		client:                    client,
+		ftcInformer:               ftcInformer,
+		eventHandlerGenerators:    []*EventHandlerGenerator{},
+		gvrMapping:                bijection.NewBijection[string, schema.GroupVersionResource](),
+		informers:                 map[string]informers.GenericInformer{},
+		informerCancelFuncs:       map[string]context.CancelFunc{},
+		eventHandlerRegistrations: map[string]map[*EventHandlerGenerator]cache.ResourceEventHandlerRegistration{},
+		lastAppliedFTCsCache:      map[string]map[*EventHandlerGenerator]*fedcorev1a1.FederatedTypeConfig{},
+		queue:                     workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter()),
+	}
+
+	ftcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { manager.enqueue(obj) },
+		UpdateFunc: func(_ interface{}, obj interface{}) { manager.enqueue(obj) },
+		DeleteFunc: func(obj interface{}) { manager.enqueue(obj) },
+	})
+
+	return manager
+}
+
+func (m *informerManager) enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Error(err, "informer-manager: Failed to enqueue FederatedTypeConfig")
+		return
+	}
+	m.queue.Add(key)
+}
+
+func (m *informerManager) worker(ctx context.Context) {
+	key, shutdown := m.queue.Get()
+	if shutdown {
+		return
+	}
+	defer m.queue.Done(key)
+
+	logger := klog.FromContext(ctx)
+
+	_, name, err := cache.SplitMetaNamespaceKey(key.(string))
+	if err != nil {
+		logger.Error(err, "Failed to process FederatedTypeConfig")
+		return
+	}
+
+	logger = logger.WithValues("ftc", name)
+	ctx = klog.NewContext(ctx, logger)
+
+	ftc, err := m.ftcInformer.Lister().Get(name)
+	if apierrors.IsNotFound(err) {
+		if err := m.processFTCDeletion(ctx, name); err != nil {
+			logger.Error(err, "Failed to process FederatedTypeConfig, will retry")
+			m.queue.AddRateLimited(key)
+			return
+		}
+		return
+	}
+	if err != nil {
+		logger.Error(err, "Failed to get FederatedTypeConfig from lister, will retry")
+		m.queue.AddRateLimited(key)
+		return
+	}
+
+	err, needReenqueue := m.processFTC(ctx, ftc)
+	if err != nil {
+		if needReenqueue {
+			logger.Error(err, "Failed to process FederatedTypeConfig, will retry")
+		} else {
+			logger.Error(err, "Failed to process FederatedTypeConfig")
+		}
+	}
+	if needReenqueue {
+		m.queue.AddRateLimited(key)
+	}
+}
+
+func (m *informerManager) processFTC(ctx context.Context, ftc *fedcorev1a1.FederatedTypeConfig) (err error, needReenqueue bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	ftcName := ftc.Name
+	apiResource := ftc.GetSourceType()
+	gvr := schemautil.APIResourceToGVR(&apiResource)
+
+	logger := klog.FromContext(ctx).WithValues("gvr", gvr.String())
```

Comment:
Should we make a util method for this pattern?

Reply:
Added an extension method to `FederatedTypeConfig` to directly return `schema.GroupVersionResource`

Reply:
I was referring to adding a KV pair into the logger and assigning the logger to the context.

Reply:
That'll be incredibly useful too

Reply:
Done